### ISSUE TITLE
fig.set_dpi() does not set the dpi correctly

### ISSUE THREAD

**Issue by ImportanceOfBeingErnest** [matplotlib/matplotlib#11227]
# Bug report

### Bug summary

Changing the figure's dpi via `fig.set_dpi()` does not have the desired effect. There are strange things happening when saving or showing the figure as detailed below.

### Bug 1 - saving

In the following a figure with dpi=200 should be saved in all 3 cases.

```
import matplotlib.pyplot as plt

get_img_size = lambda path: print(plt.imread(path).shape)

fig1, ax1 = plt.subplots(dpi=200)
ax1.plot([1,2])
fig1.savefig("fig1.png")
get_img_size("fig1.png")

fig2, ax2 = plt.subplots(dpi=100)
ax2.plot([1,2])
fig2.savefig("fig2.png", dpi=200)
get_img_size("fig2.png")

fig3, ax3 = plt.subplots(dpi=100)
ax3.plot([1,2])
fig3.set_dpi(200)
fig3.savefig("fig3.png", dpi="figure")
get_img_size("fig3.png")
```

**Qt5Agg**

The outcome using the Qt5Agg backend of this is 

```
(960, 1280, 4)
(960, 1280, 4)
(480, 640, 4)
```
i.e. in the first two cases we get the expected 200 dpi figure. In the last case, where the figure dpi is set via **`fig.set_dpi` it seems to be ignored** and the output image has a dpi of 100.

**TkAgg**  

Using the TkAgg backend, the saved figures are as expected (`(960, 1280, 4)` in all cases.) 

### Bug 2 - showing

In the following we would expect to get two identical figures shown, both with dpi=100 and 640x480 pixels large.

```
plt.close("all")

fig1, ax1 = plt.subplots(num="expected", dpi=100)
ax1.plot([1,2])

fig3, ax3 = plt.subplots(num="buggy", dpi=50)
ax3.plot([1,2])
fig3.set_dpi(100)
print(fig3.get_dpi())
fig3.savefig("fig3a.png", dpi="figure")
print(fig3.get_dpi())
plt.show()
print(fig3.get_dpi())
```

**Qt5Agg**

The output (with Qt5Agg) is the following:

```
100
100
50
```

![image](https://user-images.githubusercontent.com/23121882/39899083-43d2f03c-54ba-11e8-9642-54aa1b243134.png)

The printed dpi is correctly 100 before showing the figure. After showing, it prints 50. The shown figures have both the correct size, but the figure where dpi is set via `fig.set_dpi` has narrower linewidth and smaller fontsize. It looks exaclty like a figure that would have been expected for dpi=50, just with the wrong pixel size. 

**TkAgg**

Using TkAgg backend, the output is correctly printed as `100` in all cases, **but** the shown figures differ in size; the left (expected) figure is 640x480 pixels, while the right figure is 640x445 pixels.

![image](https://user-images.githubusercontent.com/23121882/39899772-1f1abe10-54be-11e8-99cb-f09df028fd44.png)

**Matplotlib version**

  * Operating system: Windows 8.1
  * Matplotlib version: master (2.2.2.post1088.dev0+g9ec4b95d6)
  * Matplotlib backend: Qt5Agg & TkAgg (see text)
  * Python version: 3.6.4

<hr>

Running the same with matplotlib 2.0.2 (all other versions the same) I get 
Qt5Agg, saving: `(960, 1280, 4), (960, 1280, 4), (480, 640, 4)`  - same bug as above with master.
Qt5Agg, showing `100, 100, 100` - printed dpi are all the same, **shown figures are the same**.
TkAgg, saving: `(960, 1280, 4)` in all cases. - same expected behaviour as above.
TkAgg, showing `100, 100, 100` - printed dpi are all the same, shown figures differ in same manner as above.

**Comment by jklymak** [comment]
yes exhibit A as to why setting the figure dpi *before* rendering make our lives a lot harder.  I had similar bug reports 6 months ago.

**Comment by ImportanceOfBeingErnest** [comment]
I tested everything via `python script.py`. But I observe the same when running them in Spyder via the IPython Qt console (which is how I usually work with python); I suppose that would be the same as typing `runfile(script.py)` into IPyton. Does that qualify as "interactively" or do you mean something else?

**Comment by ImportanceOfBeingErnest** [comment]
The high-res stuff might explain the problems of Bug 2 I suppose. But if there is no figure shown at all, none of this should take any effect, right?

**Comment by jklymak** [comment]
For figure: 
```
        dpi : [ *None* | scalar > 0 | 'figure' ]
            The resolution in dots per inch.  If *None*, defaults to
            :rc:`savefig.dpi`.  If 'figure', uses the figure's dpi value.
```
So I think if you don't specify the dpi it'll fall back to the default, so that is your first case.  Why it does something different for tk is beyond me...

**Comment by jklymak** [comment]
For `plt.show` the question is what do we expect to see if you have set the figure dpi to something different than your screen resolution?

**Comment by tacaswell** [comment]
No, the first two are correct because for the first one you are setting the figure DPI to 200 (which will render to the screen at 400) and save at 200, in the second case you are telling savefig to use 200 DPI (which it is).  The third case is exposing a race condition someplace between the figure state and the GUI state.

**Comment by ImportanceOfBeingErnest** [comment]
What the heck does "screen resolution" mean here? I'm not measuring anything with a ruler on my screen. But just because you are asking, I have a 1920x1080 pixel monitor with a diagonal of 27 inch, hence my screen resolution is `sqrt(1920^2+1080^2)/27. = 82 ppi`. 
Nice to know - but completely unrelated to the bug report.

**Comment by ImportanceOfBeingErnest** [comment]
> figure DPI to 200 (which will render to the screen at 400) and save at 200

This does not make any sense to me at all. Where is the number 400 coming from? And why should it save at 200 dpi, if - as the documentation states - the default rcParam is used, which is 100? 

But as I said, feel free to replace the respective line by `fig1.savefig("fig1.png", dpi="figure")`, the complete bug report stays valid.

**Comment by jklymak** [comment]
OK, so on hi-dpi my machine:

```python

import matplotlib
#matplotlib.use('tkagg')
import matplotlib.pyplot as plt

plt.close("all")

fig1, ax1 = plt.subplots(num="expected", dpi=100)
ax1.plot([1,2])

fig3, ax3 = plt.subplots(num="buggy", dpi=50)
fig3.set_dpi(100)
plt.show()

print(fig3.get_dpi())
```

prints `100` but yields the original 50 dpi...

<img width="1036" alt="buggy_and_expected_and_1__python_checkdpi2_py__python3_6__and_2__tmux_-cc__zsh_" src="https://user-images.githubusercontent.com/1562854/39903963-9b50bf72-548a-11e8-889d-a89f74518faf.png">

**Comment by tacaswell** [comment]
The high-dpi GUI frameworks need a double-resolution image for painting to the screen.  It is rather annoying, but it is the way it is (you can dig up quite a few comments from me about a year ago complaining vociferously about this, but I have moved on to acceptance).

The default value of `savefig.dpi' is 'figure'.

https://github.com/matplotlib/matplotlib/blob/9ec4b95d6ae76e30a5ff2d6e4d19ec0d15466a1a/lib/matplotlib/figure.py#L1852
https://github.com/matplotlib/matplotlib/blob/9ec4b95d6ae76e30a5ff2d6e4d19ec0d15466a1a/lib/matplotlib/rcsetup.py#L1347

which eventually falls through to backend bases

https://github.com/matplotlib/matplotlib/blob/9ec4b95d6ae76e30a5ff2d6e4d19ec0d15466a1a/lib/matplotlib/backend_bases.py#L2147-L2148

to resolve the string to a number.  Note that this looks for an optional `_original_dpi` attribute on the figure which is set
https://github.com/matplotlib/matplotlib/blob/9ec4b95d6ae76e30a5ff2d6e4d19ec0d15466a1a/lib/matplotlib/backends/backend_qt5.py#L232

and similarly in webagg.  However it is not updated when `set_dpi` is called which is likely the source of the bug.

@jklymak  Either we set the dpi at the figure level and fold it into the transform stack or we would have to pass the dpi into every single backend draw method.  We also use the dpi + figure size to sort out how big to make the gui windows.

**Comment by jklymak** [comment]
...glad you moved onto acceptance, I'm still in denial :wink:  Thanks for bearing w/ me!

**Comment by jklymak** [comment]
See #11232 for what I think is a reasonable fix that shouldn't break anything....

**Comment by BearBearCodes** [comment]
I am experiencing a similar issue on `matplotlib 3.5.2` with `python 3.10.4` running on WSL2 with Ubuntu 20.04.5 LTS.

---
tl;dr:
- `fig.set_dpi(150)` only changes the dpi of plot shown in the Jupyter notebook but not in the saved figure
- `figure.dpi` in the `matplotlibrc` file does not work at all
- `plt.subplots(dpi=150)` changes the dpi of both the plot shown in the Jupyter notebook as well as in the saved figure
---

More details:

When I run the following code in a Jupyter notebook running `IPython 8.1.1`,
```python
fig, ax = plt.subplots()
fig.set_dpi(150)
ax.plot([1, 2], [1, 2])
print(fig.get_dpi())
fig.savefig("test.png", dpi="figure")
plt.show()
```
I find that the Jupyter preview window reflects the increased DPI but the saved figure does NOT respect the changed dpi (it is still at the default 72 dpi).

Moreover, I'm finding that on my machine, the `figure.dpi` rcParam is not respected after I have modified it in my `matplotlibrc` file (located at `~/.config/matplotlib/matplotlibrc`). Other changes to options like `axes.prop_cycle` in my `matplotlibrc` file _are_ being reflected in the plots shown in the Jupyter notebook as well as in those saved via `fig.savefig()`, but changes to the `figure.dpi` option are not reflected in the plots shown/saved. Note that if I instead set the dpi via `plt.subplots(dpi=150)`, then the plots shown/saved both have a dpi of 150, as expected.

**Comment by jklymak** [comment]
Actually let's close this, since DPI handling in most of the backends has changed significantly.  @BearBearCodes if you can open a new issue with all the required information, that would be more clear.  Thanks!

**Comment by BearBearCodes** [comment]
> Actually let's close this, since DPI handling in most of the backends has changed significantly. @BearBearCodes if you can open a new issue with all the required information, that would be more clear. Thanks!

Okay I will do that. Thank you!

**Comment by BearBearCodes** [comment]
@jklymak yep I am using running both the Jupyter notebook and the standalone python script from a conda environment.
